File: backend/handlers/algorithms.go
package handlers

import (
	"encoding/json"
	"net/http"

	"github.com/cuqz/array-sorter/models"
)

var Algorithms = []models.AlgorithmInfo{
	{ID: "bubble", Name: "Bubble Sort", Category: "comparison", TimeComplexity: "O(n\u00b2)", Stable: true, Description: "Classic adjacent-swap sort. Slow but every CS student writes it once."},
	{ID: "selection", Name: "Selection Sort", Category: "comparison", TimeComplexity: "O(n\u00b2)", Stable: false, Description: "Picks the smallest element each pass. Worst of both worlds on comparisons and swaps."},
	{ID: "insertion", Name: "Insertion Sort", Category: "comparison", TimeComplexity: "O(n\u00b2)", Stable: true, Description: "Good for small or nearly-sorted arrays. Builds the sorted array one element at a time."},
	{ID: "shell", Name: "Shell Sort", Category: "comparison", TimeComplexity: "O(n log\u00b2 n)", Stable: false, Description: "Insertion sort with a decreasing gap. First O(n log n) sort to be discovered."},
	{ID: "quick", Name: "Quick Sort", Category: "comparison", TimeComplexity: "O(n log n)", Stable: false, Description: "Pivot, partition, recurse. Fast in practice unless data is already sorted with a bad pivot."},
	{ID: "merge", Name: "Merge Sort", Category: "comparison", TimeComplexity: "O(n log n)", Stable: true, Description: "Divide-and-conquer. Splits to single elements, merges back up in sorted order."},
	{ID: "heap", Name: "Heap Sort", Category: "comparison", TimeComplexity: "O(n log n)", Stable: false, Description: "Builds a heap, pops the max repeatedly. Consistent O(n log n) but not stable."},
	{ID: "tim", Name: "Tim Sort", Category: "comparison", TimeComplexity: "O(n log n)", Stable: true, Description: "Hybrid of merge and insertion sort. Powers Python's sort and V8's Array.sort()."},
	{ID: "cocktail", Name: "Cocktail Shaker Sort", Category: "comparison", TimeComplexity: "O(n\u00b2)", Stable: true, Description: "Bubble sort but bidirectional. Shakes left-to-right then right-to-left each pass."},
	{ID: "comb", Name: "Comb Sort", Category: "comparison", TimeComplexity: "O(n\u00b2)", Stable: false, Description: "Bubble sort with a shrinking gap. Shrink factor of 1.3 is empirically best."},
	{ID: "gnome", Name: "Gnome Sort", Category: "comparison", TimeComplexity: "O(n\u00b2)", Stable: true, Description: "Like insertion sort if it were drunk. Swaps backward until in position."},
	{ID: "counting", Name: "Counting Sort", Category: "linear", TimeComplexity: "O(n+k)", Stable: true, Description: "Integer-only. Counts occurrences and reconstructs. Falls back to native sort if range too large."},
	{ID: "radix", Name: "Radix Sort", Category: "linear", TimeComplexity: "O(nk)", Stable: true, Description: "Sorts digit-by-digit from LSD to MSD. Counting sort as a subroutine."},
	{ID: "bucket", Name: "Bucket Sort", Category: "linear", TimeComplexity: "O(n + k)", Stable: true, Description: "Distributes into buckets, sorts each bucket, concatenates. Good for uniform distributions."},
	{ID: "bogo", Name: "Bogo Sort", Category: "comparison", TimeComplexity: "O((n+1)!)", Stable: false, Description: "Shuffles randomly until sorted. Included because it is funny. Gives up after 1000 tries."},
}

func AlgorithmsHandler(w http.ResponseWriter, r *http.Request) {
	resp := struct {
		Algorithms []models.AlgorithmInfo `json:"algorithms"`
	}{Algorithms}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(resp)
}

File: frontend/src/utils/algorithms.ts
export function getDisplayName(id: string): string {
  const names: Record<string, string> = {
    bubble: 'Bubble Sort', selection: 'Selection Sort', insertion: 'Insertion Sort',
    shell: 'Shell Sort', quick: 'Quick Sort', merge: 'Merge Sort',
    heap: 'Heap Sort', tim: 'Tim Sort', cocktail: 'Cocktail Sort',
    comb: 'Comb Sort', gnome: 'Gnome Sort', counting: 'Counting Sort',
    radix: 'Radix Sort', bucket: 'Bucket Sort', bogo: 'Bogo Sort', custom: 'Custom',
  };
  return names[id] || id;
}

export const STABLE_ALGOS = new Set(['bubble','insertion','merge','tim','cocktail','gnome','counting','radix','bucket']);

export function isStable(id: string): boolean {
  return STABLE_ALGOS.has(id);
}

export function getComplexity(id: string): string {
  const c: Record<string, string> = {
    bubble: 'O(n²)', selection: 'O(n²)', insertion: 'O(n²)', shell: 'O(n log² n)',
    quick: 'O(n log n)', merge: 'O(n log n)', heap: 'O(n log n)', tim: 'O(n log n)',
    cocktail: 'O(n²)', comb: 'O(n²)', gnome: 'O(n²)',
    counting: 'O(n + k)', radix: 'O(nk)', bucket: 'O(n + k)', bogo: 'O((n+1)!)',
    custom: '—',
  };
  return c[id] || '—';
}
